### PR TITLE
feat(typescript): initiate typescript conversion

### PR DESCRIPTION
**Pull request by 0ff1ane**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+## [1.2.0](https://github.com/MichaelHirn/ts-template/compare/v1.1.0...v1.2.0) (2020-05-25)
+
+
```
Comment (MichaelHirn) [inline CHANGELOG.md]: This file should not be committed. This is the CHANGELOG for ts-template

File: README.example.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# [Project Name]
+
+## Development
```
Comment (MichaelHirn) [inline README.example.md]: This file should not be committed. This is an example README from ts-template

File: etc/ts-template.api.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
+{
+  "metadata": {
+    "toolPackage": "@microsoft/api-extractor",
+    "toolVersion": "7.8.4",
```
Comment (MichaelHirn) [inline etc/ts-template.api.json]: This file should not be committed. This is the `ts-template` api extract.

File: etc/ts-template.api.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+## API Report File for "ts-template"
+
+> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).
```
Comment (MichaelHirn) [inline etc/ts-template.api.md]: This file should not be committed. This is the `ts-template` api doc.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { Cache } from './cache'
+import { Client } from './client'
+import { ClientLowLevel } from './clientLowLevel'
+
+export {
+  Client,
+  ClientLowLevel,
+  Cache
+}
```
Comment (MichaelHirn) [inline src/index.ts]: you can write 
```ts
export { Cache } from './cache'
export { Client } from './client'
export { ClientLowLevel } from './clientLowLevel'
```

File: src/debug.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+import * as debug from 'debug'
+
+module.exports = debug('authless-client')
```
Comment (MichaelHirn) [inline src/debug.ts]: does this even work?

**Review by MichaelHirn — CHANGES_REQUESTED** [review]
did you run this command?

```
find ./src -name "*.js" -exec sh -c 'git mv "$0" "${0%.js}.ts"' {} \;
```

If you change checked in files e.g. `src/example.js` -> `src/example.ts` with `git mv` instead of `mv` git can track the change of the file and not track two different files. Here the .js and .ts files are separate and it's difficult to see the change of the file now.

File: src/debug.ts (hunk 6 above)
Comment (MichaelHirn) [reply src/debug.ts]: this should be `export default debug('authless-client')`

File: src/cache.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import * as VError from 'verror'
+const debug = require('./debug').extend('cache')
```
Comment (MichaelHirn) [inline src/cache.ts]: no `require` - once `debug.ts` calls export this won't work anymore

**Review by MichaelHirn — CHANGES_REQUESTED** [review]
the command `npm run docs` fails - needs fixing.

File: src/cache.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+}
+class Cache {
+  name: string
+  debug: any
+  existsFn: (url: string) => Promise<boolean>
+  storeFn: (url: string, data: object) => object
+  retrieveFn: (url: string) => object
+  removeFn: (url: string) => object
+
+  static isValidConfig (x: any): x is Config {
```
Comment (MichaelHirn) [inline src/cache.ts:21]: why `(x: any): x is Config` and not `(x: Config): boolean`?

File: src/client.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
+import * as VError from 'verror'
+// import * as check from 'check-types'
```
Comment (MichaelHirn) [inline src/client.ts]: no comment abuse

File: src/cache.ts (hunk 8 above)
Comment (0ff1ane) [reply src/cache.ts:21]: this is a type-guard, if we use this in a `if` condition, typescript will assume the type of the parameter to be `Config` inside the block of that `if` condition

File: src/client.ts (hunk 9 above)
Comment (0ff1ane) [reply src/client.ts]: ah, :)

**Review by MichaelHirn — APPROVED** [review]
LGTM now 👍 

will merge

**Pull request merged**